stellargraph/layer/watch_your_step.py

```diff
@@ -0,0 +1,190 @@
+    to use for the node embeddings. Note, that the embedding dimension should be an even number or else will be
+    rounded down to nearest even number.
+
+    Args:
+        generator (AdjacencyPowerGenerator): the generator
+        num_walks (int): the number of random walks starting at each node to use when calculating the expected random
+            walks. Defaults to `80` as this value was found to perform well by the authors of the paper.
+        embedding dimension (int): the dimension to use for the node embeddings. Defaults to `64`
+            as this value was found to perform well by the authors of the paper.
+        attention_initializer (str or func, optional): The initialiser to use for the attention weights.
+        attention_regularizer (str or func, optional): The regulariser to use for the attention weights.
+        attention_constraint (str or func, optional): The constraint to use for the attention weights.
+    """
+
+    def __init__(
+        self,
+        generator,
+        num_walks=80,
+        embedding_dimension=64,
+        attention_initializer="glorot_uniform",
+        attention_regularizer=None,
+        attention_constraint=None,
+    ):
+
+        if not isinstance(generator, AdjacencyPowerGenerator):
+            raise TypeError(
+                "generator should be an instance of AdjacencyPowerGenerator."
+            )
+
+        if not isinstance(num_walks, int):
+            raise TypeError("num_walks should be an int.")
+
+        if num_walks <= 0:
+            raise ValueError("num_walks should be a positive int.")
+
+        self.num_walks = num_walks
+        self.num_powers = generator.num_powers
+        self.n_nodes = int(generator.Aadj_T.shape[0])
+
+        if embedding_dimension % 2:
+            embedding_dimension = embedding_dimension - 1
+            warnings.warn(
+                "embedding_dimension must be even. embedding_dimension will be rounded down to the nearest even number."
+            )
+        self.embedding_dimension = embedding_dimension
+        self.attention_regularizer = attention_regularizer
+        self.attention_initializer = attention_initializer
+        self.attention_constraint = attention_constraint
+
+    def build(self):
+        """
+        This function builds the layers for a keras model.
+
+        returns:
+            A tuple of (inputs, outputs) to use with a keras model.
+        """
+
+        input_rows = Input(batch_shape=(None,), name="row_node_ids", dtype="int64")
+        input_powers = Input(batch_shape=(None, self.num_powers, self.n_nodes))
+
+        left_embedding = Embedding(
+            self.n_nodes,
+            int(self.embedding_dimension / 2),
+            input_length=None,
+            name="WATCH_YOUR_STEP_LEFT_EMBEDDINGS",
+        )
+
+        vectors_left = left_embedding(input_rows)
+        outer_product = Dense(self.n_nodes, name="WATCH_YOUR_STEP_RIGHT_EMBEDDINGS")(
```

Comment:
Nice 👍 

Could you add a comment about how this corresponds to the paper, because it's a significant departure from how it is phrased there (but, hopefully, not in its behaviour)?

Reply:
I've added in a comment - let me know if it needs more clarity :) 

weirdly changing to a `Dense` layer didn't speed things up. Maybe `tensorflow` had already optimized around the weird lambda layers and lookups